### ISSUE TITLE
Clarifications of Aggregation

### ISSUE THREAD

**Issue by pmartinou** [openconfig/reference#77]
While trying to understand how to use aggregation, some points remained obscure.
The text below is a proposal for clarification (but is my understanding correct ?)

A same node cannot belong to two different Aggregates. For example, in the sample tree of section 2.4.3, it would be impossible to declare both ChildA and ChildA3 as Aggregates.

An Aggregate is defined as a set of siblings. Nodes from different Containers cannot be Aggregated together (this would require an extra specification of the form of returned objects).
For example, with the sample tree of section 2.4.3, one may specify that /childA/childA3 is an Aggregate, not that we have an Aggregate composed of leafA1 and leafA31.
All the elements of an Aggregate are always returned together. This implies that when a child element changes, the whole aggregate is always sent to the client: for example, with the same sample tree, for a Client having a STREAM/onChange subscription covering it, a value change of leafA31 would result in a SubscribeResponse with a Notification containing childA3, not just leafA31.

**Comment by robshakir** [comment]
We have an action currently to clarify the use of aggregation in gNMI. The intention is that this can only be done at levels that are specified within the schema to allow for aggregation, so there is some case whereby the *schema* will be what defines the set of nodes that are to be aggregated together, vs. something else.

In YANG, we're currently discussing that a subtree is an aggregate, and that it can be serialised into a protobuf, rather than into JSON. If a container were marked as aggregatable, then it would imply that all children (and their children) would be affected by such aggregation.

**Comment by pmartinou** [comment]
First (later points to be discussed in a next post), why an aggregate would necessarily be serialized into a protobuf: couldn't the corresponding subtree be serialized as a JSON object ?

**Comment by robshakir** [comment]
Our analysis of serialising to JSON was that the development efforts required to add this to the collector, and the use cases, seemed relatively minimal. Aggregation necessarily gives larger on the wire updates than <path,value> updates (or certainly, less clearly fragmentable updates) -- our implementation experience points towards wanting more efficient on the wire encoding in such cases.

**Comment by pmartinou** [comment]
Yes it is clear that the definition of aggregates must be done at schema level.
I had in mind that an aggregate could be defined as a set of siblings, not necessarily taking into account all children of a container, but it is certainly better to define it at container level: simpler to understand and specify.

As an example of aggregation, the following use case can be considered:

The input signal level, output signal level, status, etc. of an amplifier for both directions (AB, BA) are always returned together by the hardware.
In such a model, the values for one Direction are grouped as an "optical-amp-state"  (see definition below), with an instance of it for each Direction, each being the "state" attribute of an Amplifier object.

While it is true that the values for both directions are acquired at the same time from the hardware, aggregating them would be more complex, as each "state" belongs to a different container. 
It makes therefore more sense to tag the values for a single Direction as an Aggregate, and that fits well in the proposed scheme to define Aggregation at container level.

```
grouping optical-amp-state {	<<< Instances of optical-amp-state will be Aggregates
  container input-power {
    uses optical-power;
  }
  container output-power {
    uses optical-power;
  }
  leaf status-flag {
    type enumeration {
      enum raised;
      enum cleared;
      enum not-applicable;
    }
  }
  // Other attributes
}

container ab-amplifier {
  container state {
    config false;
    uses optical-amp-state;
  }
  // other attributes 
}
container ba-amplifier {
  container state {
    config false;
    uses optical-amp-state;
  }
  // other attributes 
}
```
